pilot/pkg/controllers/ipallocate/ipallocate.go

```diff
@@ -0,0 +1,407 @@
+		log.Debugf("reassigned %s/%s due to IP Address conflict", resolveMe.Namespace, resolveMe.Name)
+		patch, err := c.statusPatchForAddresses(resolveMe, true)
+		if err != nil {
+			errs = errors.Join(errs, err)
+		}
+
+		if patch == nil {
+			errs = errors.Join(errs, fmt.Errorf("this should not occur but patch was empty on a forced reassignment"))
+		}
+
+		_, err = c.serviceEntryClient.PatchStatus(resolveMe.Name, resolveMe.Namespace, types.MergePatchType, patch)
+		if err != nil {
+			errs = errors.Join(errs, err)
+		}
+	}
+
+	return errs
+}
+
+func allAddresses(se *networkingv1alpha3.ServiceEntry) ([]netip.Addr, []netip.Addr) {
+	if se == nil {
+		return nil, nil
+	}
+	autoAssigned := autoallocate.GetV2AddressesFromServiceEntry(se)
+	userAssigned := []netip.Addr{}
+
+	for _, a := range se.Spec.Addresses {
+		parsed, err := netip.ParseAddr(a)
+		if err != nil {
+			// maybe log
+			continue
+		}
+		userAssigned = append(userAssigned, parsed)
+	}
+	return autoAssigned, userAssigned
+}
+
+func (c *IPAllocate) nextAddresses(owner types.NamespacedName) []netip.Addr {
+	results := []netip.Addr{}
+	v4address, err := c.v4allocator.allocateNext(owner)
+	if err != nil {
+		log.Error("failed to allocate v4 address")
+	} else {
+		results = append(results, v4address)
+	}
+	v6address, err := c.v6allocator.allocateNext(owner)
+	if err != nil {
+		log.Error("failed to allocate v6 address")
+	} else {
+		results = append(results, v6address)
+	}
+	return results
+}
+
+func (c *IPAllocate) inOurRange(a netip.Addr) bool {
+	a = a.Unmap()
+	if a.Is6() {
+		return c.v6allocator.prefix.Contains(a)
+	}
+
+	return c.v4allocator.prefix.Contains(a)
+}
+
+func (c *IPAllocate) markUsedOrQueueConflict(a netip.Addr, owner types.NamespacedName) {
+	if a.Is6() {
+		if used, usedByOwner := c.v6allocator.markUsed(a, owner); used && !usedByOwner {
+			c.queue.Add(conflictDetectedEvent{
+				conflictingResourceIdentifier: owner,
+				conflictingAddress:            a,
+			})
+		}
+	} else {
+		if used, usedByOwner := c.v4allocator.markUsed(a, owner); used && !usedByOwner {
+			c.queue.Add(conflictDetectedEvent{
+				conflictingResourceIdentifier: owner,
+				conflictingAddress:            a,
+			})
+		}
+	}
+}
+
+func (c *IPAllocate) statusPatchForAddresses(se *networkingv1alpha3.ServiceEntry, forcedReassign bool) ([]byte, error) {
+	if se == nil {
+		return nil, nil
+	}
+	orig, err := json.Marshal(se)
+	if err != nil {
+		return nil, err
+	}
+
+	if forcedReassign {
+		// TODO: this is a kludge and incorrect, we should not overwrite status in the final commit
+		// initialize status field
+		se.Status = v1alpha1.IstioStatus{}
+	}
+	foundAddresses := autoallocate.GetV2AddressesFromServiceEntry(se)
+	if len(foundAddresses) > 0 && !forcedReassign {
+
+		// this is likely a noop, but just to be safe we should check and potentially resolve conflict
+		for _, a := range foundAddresses {
+			c.markUsedOrQueueConflict(a, types.NamespacedName{
```

Comment:
is this duplicate while calling from `populateControllerDatastructures`

Reply:
I did clean up a lot of this. It's probably not perfect still but it's better than it was I think